File: BinaryTree/BinaryTree/BinaryTreeNode.cpp
#include "BinaryTreeNode.h"



BinaryTreeNode::BinaryTreeNode()
{
	data = 0;
	left = nullptr;
	right = nullptr;
}


BinaryTreeNode::~BinaryTreeNode()
{
	delete left;
	delete right;
	
}

BinaryTreeNode::BinaryTreeNode(int newData)
{
	this->data = newData;
	this->left = nullptr;
	this->right = nullptr;
}

BinaryTreeNode::BinaryTreeNode(const BinaryTreeNode & copyNode)
{
	this->data = copyNode.data;
	this->right = copyNode.right;
	this->left = copyNode.left;
}

BinaryTreeNode * BinaryTreeNode::getLeft()
{
	return this->left;
}

BinaryTreeNode * BinaryTreeNode::getRight()
{
	return this->right;
}

BinaryTreeNode * BinaryTreeNode::getParent()
{
	return parent;
}

int BinaryTreeNode::getData()
{
	return this->data;
}

void BinaryTreeNode::setLeft(BinaryTreeNode * newLeft)
{
	this->left = newLeft;
}

void BinaryTreeNode::setRight(BinaryTreeNode * newRight)
{
	this->right = newRight;
}

void BinaryTreeNode::setParent(BinaryTreeNode * newParent)
{
	this->parent = newParent;
}

void BinaryTreeNode::setData(int newData)
{
	this->data = newData;
}

int BinaryTreeNode::getDepth(BinaryTreeNode* node)
{
	int depth = 0;
	while (node != nullptr) {
		node = node->getParent();
		depth++;
	}
	return depth;
}

File: Java Practice/Data Structure/BinaryTreeGenerics/src/BinaryTreeNode.java

public class BinaryTreeNode<T extends Comparable<T>>  {
	
	private T data;
	private BinaryTreeNode<T> left;
	private BinaryTreeNode<T> right;
	
	public BinaryTreeNode(T newData) {
		data = newData;			
		left = null;
		right = null;
	}
	
	public BinaryTreeNode(BinaryTreeNode<T> copyNode) {
		this.data = copyNode.getData();
		this.left = new BinaryTreeNode<T>(copyNode.getLeft().getData());
		this.right = new BinaryTreeNode<T>(copyNode.getRight().getData());
		
	}
	
	public void setData(T newData) {
		data = newData;
	}
	
	public void setLeft (BinaryTreeNode<T> newLeft) {
		this.left =  newLeft;
	}
	
	public void setRight(BinaryTreeNode<T> newRight) {
		this.right = newRight;
	}
	
	public T getData() {
		return data;
	}
	
	public BinaryTreeNode<T> clone(){
		return  this;
	}
	
	public BinaryTreeNode<T> getRight(){
		return right;
	}
	
	public BinaryTreeNode<T> getLeft(){
		return left;
	}
	
	public BinaryTreeNode<T> getMostLeft(){
		if(this.getLeft() != null) {
			return getMostLeft(this.getLeft());
		}
		else {
			return null;
		}
	}
	
	public BinaryTreeNode<T> getMostLeft(BinaryTreeNode<T> node){
		if(node.getLeft() != null) {
			return getMostLeft(node.getLeft());
		}
		else {			
			return node;
		}
	}
	
	public BinaryTreeNode<T>getMostRight(){
		if(this.getRight() != null) {
			return getMostRight(this.getRight());
		}
		else {
			return null;
		}
	}
	
	public BinaryTreeNode<T> getMostRight(BinaryTreeNode<T> node){
		if(node.getRight() != null) {
			return getMostRight(node.getRight());
		}
		return node;
	}
}
